Radix<T: 'static>, Stack<T: 'static>?

I am porting `rsdb` to crossbeam-epoch (https://github.com/jeehoonkang/rsdb/tree/crossbeam-epoch), and currently, it doesn't compile for lifetime reasons.  In crossbeam-epoch, `fn defer_drop(obj)` requires that `obj` is a `Ptr` to a ` **'static type**:

```rust
unsafe fn defer_drop<T: Send + 'static>(self, ptr: Ptr<'scope, T>);
```

This is because we don't know actually when it will be dropped.  Similarly to `defer_free()`.

However, `rsdb`'s stack and radix doesn't guarantee `T: 'static`.  But it seems all instantiations satisfy this condition.

Do you think it is okay to add the constraint `T: 'static` to both `Radix` and `Stack`?


FYI, here is the compilation error log:

```
   Compiling rsdb v0.11.2 (file:///home/jeehoonkang/Works/rsdb)
error[E0310]: the parameter type `T` may not live long enough
   --> src/ds/radix.rs:126:31
    |
126 |                         scope.defer_drop(old);
    |                               ^^^^^^^^^^
    |
    = help: consider adding an explicit lifetime bound `T: 'static`...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/ds/radix.rs:126:31
    |
126 |                         scope.defer_drop(old);
    |                               ^^^^^^^^^^

error[E0310]: the parameter type `T` may not live long enough
   --> src/ds/radix.rs:149:28
    |
149 |             unsafe { scope.defer_drop(old) };
    |                            ^^^^^^^^^^
    |
    = help: consider adding an explicit lifetime bound `T: 'static`...
note: ...so that the type `T` will meet its required lifetime bounds
   --> src/ds/radix.rs:149:28
    |
149 |             unsafe { scope.defer_drop(old) };
    |                            ^^^^^^^^^^

error[E0310]: the parameter type `T` may not live long enough
   --> src/ds/stack.rs:183:32
    |
183 |                 unsafe { scope.defer_drop(old) };
    |                                ^^^^^^^^^^
    |
    = help: consider adding an explicit lifetime bound `T: 'static`...
note: ...so that the type `ds::stack::Node<T>` will meet its required lifetime bounds
   --> src/ds/stack.rs:183:32
    |
183 |                 unsafe { scope.defer_drop(old) };
    |                                ^^^^^^^^^^

error[E0310]: the parameter type `T` may not live long enough
   --> src/ds/stack.rs:188:32
    |
188 |                 unsafe { scope.defer_drop(new) };
    |                                ^^^^^^^^^^
    |
    = help: consider adding an explicit lifetime bound `T: 'static`...
note: ...so that the type `ds::stack::Node<T>` will meet its required lifetime bounds
   --> src/ds/stack.rs:188:32
    |
188 |                 unsafe { scope.defer_drop(new) };
    |                                ^^^^^^^^^^

error: aborting due to 4 previous errors

error: Could not compile `rsdb`.

To learn more, run the command again with --verbose.
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "rsdb"
-version = "0.11.2"
+version = "0.12.0"
 authors = ["Tyler Neely <t@jujit.su>"]
 description = "a flash-sympathetic persistent lock-free B+ tree, pagecache, and log"
 license = "Apache-2.0"

--- a/src/ds/radix.rs
+++ b/src/ds/radix.rs
@@ -17,12 +17,12 @@ fn split_fanout(i: usize) -> (usize, usize) {
     (first, rem)
 }
 
-struct Node<T> {
+struct Node<T: Send + 'static> {
     inner: Atomic<T>,
     children: Vec<Atomic<Node<T>>>,
 }
 
-impl<T> Default for Node<T> {
+impl<T: Send + 'static> Default for Node<T> {
     fn default() -> Node<T> {
         let children = rep_no_copy!(Atomic::null(); FANOUT);
         Node {
@@ -32,7 +32,7 @@ impl<T> Default for Node<T> {
     }
 }
 
-impl<T> Drop for Node<T> {
+impl<T: Send + 'static> Drop for Node<T> {
     fn drop(&mut self) {
         unsafe {
             pin(|scope| {
@@ -57,13 +57,13 @@ impl<T> Drop for Node<T> {
 
 /// A simple lock-free radix tree.
 pub struct Radix<T>
-    where T: Send + Sync
+    where T: 'static + Send + Sync
 {
     head: Atomic<Node<T>>,
 }
 
 impl<T> Default for Radix<T>
-    where T: Send + Sync
+    where T: 'static + Send + Sync
 {
     fn default() -> Radix<T> {
         let head = Owned::new(Node::default());
@@ -74,7 +74,7 @@ impl<T> Default for Radix<T>
 }
 
 impl<T> Drop for Radix<T>
-    where T: Send + Sync
+    where T: 'static + Send + Sync
 {
     fn drop(&mut self) {
         unsafe {
@@ -87,7 +87,7 @@ impl<T> Drop for Radix<T>
 }
 
 impl<T> Radix<T>
-    where T: Send + Sync
+    where T: 'static + Send + Sync
 {
     /// Try to create a new item in the tree.
     pub fn insert(&self, pid: PageID, item: T) -> Result<(), ()> {
@@ -155,7 +155,7 @@ impl<T> Radix<T>
 }
 
 #[inline(always)]
-fn traverse<'s, T>(
+fn traverse<'s, T: 'static + Send>(
     ptr: Ptr<'s, Node<T>>,
     pid: PageID,
     create_intermediate: bool,

--- a/src/ds/stack.rs
+++ b/src/ds/stack.rs
@@ -9,12 +9,12 @@ use coco::epoch::{Atomic, Owned, Ptr, Scope, pin, unprotected};
 use test_fail;
 
 #[derive(Debug)]
-pub struct Node<T: Send> {
+pub struct Node<T: Send + 'static> {
     inner: T,
     next: Atomic<Node<T>>,
 }
 
-impl<T: Send> Drop for Node<T> {
+impl<T: Send + 'static> Drop for Node<T> {
     fn drop(&mut self) {
         unsafe {
             unprotected(|scope| {
@@ -29,11 +29,11 @@ impl<T: Send> Drop for Node<T> {
 
 /// A simple lock-free stack, with the ability to atomically
 /// append or entirely swap-out entries.
-pub struct Stack<T: Send> {
+pub struct Stack<T: Send + 'static> {
     head: Atomic<Node<T>>,
 }
 
-impl<T: Send> Default for Stack<T> {
+impl<T: Send + 'static> Default for Stack<T> {
     fn default() -> Stack<T> {
         Stack {
             head: Atomic::null(),
@@ -41,7 +41,7 @@ impl<T: Send> Default for Stack<T> {
     }
 }
 
-impl<T: Send> Drop for Stack<T> {
+impl<T: Send + 'static> Drop for Stack<T> {
     fn drop(&mut self) {
         unsafe {
             unprotected(|scope| {
@@ -55,7 +55,7 @@ impl<T: Send> Drop for Stack<T> {
 }
 
 impl<T> Debug for Stack<T>
-    where T: Debug + Send + Sync
+    where T: Debug + Send + 'static + Sync
 {
     fn fmt(&self, formatter: &mut fmt::Formatter) -> Result<(), fmt::Error> {
         pin(|scope| {
@@ -78,20 +78,20 @@ impl<T> Debug for Stack<T>
     }
 }
 
-impl<T: Send> Deref for Node<T> {
+impl<T: Send + 'static> Deref for Node<T> {
     type Target = T;
     fn deref(&self) -> &T {
         &self.inner
     }
 }
 
-impl<T: Send> Node<T> {
+impl<T: Send + 'static> Node<T> {
     pub fn next<'s>(&self, scope: &'s Scope) -> Ptr<'s, Node<T>> {
         self.next.load(SeqCst, scope)
     }
 }
 
-impl<T: Send> Stack<T> {
+impl<T: Send + 'static> Stack<T> {
     /// Add an item to the stack, spinning until successful.
     pub fn push(&self, inner: T) {
         let node = Owned::new(Node {
@@ -199,14 +199,14 @@ impl<T: Send> Stack<T> {
 }
 
 pub struct StackIter<'a, T>
-    where T: 'a + Send + Sync
+    where T: 'a + Send + 'static + Sync
 {
     inner: Ptr<'a, Node<T>>,
     scope: &'a Scope,
 }
 
 impl<'a, T> StackIter<'a, T>
-    where T: 'a + Send + Sync
+    where T: 'a + Send + 'static + Sync
 {
     pub fn from_ptr<'b>(ptr: Ptr<'b, Node<T>>, scope: &'b Scope) -> StackIter<'b, T> {
         StackIter {
@@ -217,7 +217,7 @@ impl<'a, T> StackIter<'a, T>
 }
 
 impl<'a, T> Iterator for StackIter<'a, T>
-    where T: Send + Sync
+    where T: Send + 'static + Sync
 {
     type Item = &'a T;
     fn next(&mut self) -> Option<Self::Item> {
@@ -234,7 +234,7 @@ impl<'a, T> Iterator for StackIter<'a, T>
 }
 
 pub fn node_from_frag_vec<T>(from: Vec<T>) -> Owned<Node<T>>
-    where T: Send + Sync
+    where T: Send + 'static + Sync
 {
     let mut last = None;
 

--- a/src/page/mod.rs
+++ b/src/page/mod.rs
@@ -57,14 +57,14 @@ pub enum CacheEntry<M: Send + Sync> {
 /// `PageFrag` used for applying updates atomically to shared pages.
 #[derive(Debug, Clone, Copy, PartialEq)]
 pub struct CasKey<P>
-    where P: Send + Sync
+    where P: 'static + Send + Sync
 {
     ptr: *const ds::stack::Node<CacheEntry<P>>,
     tag: usize,
 }
 
 impl<'s, P> From<Ptr<'s, ds::stack::Node<CacheEntry<P>>>> for CasKey<P>
-    where P: Send + Sync
+    where P: 'static + Send + Sync
 {
     fn from(ptr: Ptr<'s, ds::stack::Node<CacheEntry<P>>>) -> CasKey<P> {
         CasKey {
@@ -75,7 +75,7 @@ impl<'s, P> From<Ptr<'s, ds::stack::Node<CacheEntry<P>>>> for CasKey<P>
 }
 
 impl<'s, P> Into<Ptr<'s, ds::stack::Node<CacheEntry<P>>>> for CasKey<P>
-    where P: Send + Sync
+    where P: 'static + Send + Sync
 {
     fn into(self) -> Ptr<'s, ds::stack::Node<CacheEntry<P>>> {
         unsafe { Ptr::from_raw(self.ptr).with_tag(self.tag) }

--- a/src/page/page_cache.rs
+++ b/src/page/page_cache.rs
@@ -68,7 +68,7 @@ use super::*;
 pub struct PageCache<PM, P, R>
     where PM: Materializer<PageFrag = P, Recovery = R>,
           PM: Send + Sync,
-          P: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
+          P: 'static + Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
           R: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send
 {
     t: PM,
@@ -84,7 +84,7 @@ pub struct PageCache<PM, P, R>
 impl<PM, P, R> Drop for PageCache<PM, P, R>
     where PM: Materializer<PageFrag = P, Recovery = R>,
           PM: Send + Sync,
-          P: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
+          P: 'static + Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
           R: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send
 {
     fn drop(&mut self) {
@@ -97,7 +97,7 @@ impl<PM, P, R> Drop for PageCache<PM, P, R>
 unsafe impl<PM, P, R> Send for PageCache<PM, P, R>
     where PM: Materializer<PageFrag = P, Recovery = R>,
           PM: Send + Sync,
-          P: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
+          P: 'static + Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
           R: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send
 {
 }
@@ -105,7 +105,7 @@ unsafe impl<PM, P, R> Send for PageCache<PM, P, R>
 unsafe impl<PM, P, R> Sync for PageCache<PM, P, R>
     where PM: Materializer<PageFrag = P, Recovery = R>,
           PM: Send + Sync,
-          P: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
+          P: 'static + Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
           R: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send
 {
 }
@@ -113,7 +113,7 @@ unsafe impl<PM, P, R> Sync for PageCache<PM, P, R>
 impl<PM, P, R> Debug for PageCache<PM, P, R>
     where PM: Materializer<PageFrag = P, Recovery = R>,
           PM: Send + Sync,
-          P: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
+          P: 'static + Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
           R: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send
 {
     fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
@@ -128,7 +128,7 @@ impl<PM, P, R> Debug for PageCache<PM, P, R>
 impl<PM, P, R> PageCache<PM, P, R>
     where PM: Materializer<PageFrag = P, Recovery = R>,
           PM: Send + Sync,
-          P: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
+          P: 'static + Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send + Sync,
           R: Debug + PartialEq + Clone + Serialize + DeserializeOwned + Send
 {
     /// Instantiate a new `PageCache`.

